compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -3162,7 +3162,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         };
         if self.can_eq(self.param_env, expected_ty, ty) {
             err.span_suggestion_short(
-                stmt.span.with_lo(tail_expr.span.hi()),
+                stmt.span
+                    .with_lo(tail_expr.span.hi())
+                    .with_hi(tail_expr.span.hi() + rustc_span::BytePos(1)),
```

Comment:
We should generally avoid manual offsets in spans. The test contains a macro, so I suppose the bug happens when both spans are from different expansions, doesn't it ? Can it be solved by using find_ancestor_inside?

Reply:
Thank you for the comment, it made me notice that there was a deeper problem. I left a comment explaining it [on the issue here](https://github.com/rust-lang/rust/issues/114251#issuecomment-1863112127).

However there is also another problem that I noticed as a result of this comment too which is caused by spans from different expansions, it suggests that you remove the semi-colon from inside the macro, even when the trailing semi-colon is outside the macro. I will try to address this too.

Reply:
I solved this, and also fixed the issue with macro spans not including their trailing semi-colon, but as I was able to fix this without the need for the macro span fixes I did not include them in this pull request. I will ask in the zulip if the fixes for macro spans are wanted, and if so open a separate PR for them.